### PR TITLE
voucher checkout

### PR DESCRIPTION
**Pull request by LukaszWiktor**
android-voucher-checkout
- added custom attributes for icons:
  - voucherIcon
  - validVoucherIcon
  - invalidVoucherIcon
- added animations
  - res/anim/valid.xml (zoom in - zoom out)
  - res/anim/invalid.xml (shake)
- added styles:
  - VoucherCodeLabel
  - VoucherCodeEditText
  - VoucherValidateButton
- prepared for release as 0.3.0

### CODE REVIEW

File: library/src/main/java/pl/rspective/voucherify/android/view/VoucherCheckoutView.java [hunk 1]
```diff
@@ -66,38 +85,75 @@ public void onClick(View v) {
...
+        }
+        invalidVoucherIcon = a.getDrawable(R.styleable.VoucherCheckoutView_invalidVoucherIcon);
+        if (invalidVoucherIcon == null) {
+            invalidVoucherIcon = ContextCompat.getDrawable(context, R.drawable.ic_error_white_24dp);
+        }
+
         a.recycle();
     }
 
+    public static Drawable getTintedDrawable(Context ctx, @DrawableRes int drawableId, @ColorRes int color) {
```
Comment (polok) [inline library/src/main/java/pl/rspective/voucherify/android/view/VoucherCheckoutView.java]: Please, move this method to some util class

File: library/src/main/java/pl/rspective/voucherify/android/view/VoucherCheckoutView.java [hunk 2]
```diff
@@ -66,38 +85,75 @@ public void onClick(View v) {
...
         String voucherCode = voucherCodeEditText.getText().toString();
 
-        voucherifyClient.voucher().async().validate(voucherCode, new VoucherifyCallback<VoucherResponse, RetrofitError>() {
+        voucherifyClient.voucher().async().validate(voucherCode, new VoucherifyCallback<VoucherResponse, VoucherifyError>() {
             @Override
             public void onSuccess(final VoucherResponse result) {
                 if (onValidatedListener != null) {
                     if (result.isValid()) {
+                        voucherCodeEditText.setCompoundDrawablesWithIntrinsicBounds(null, null, validVoucherIcon, null);
+                        voucherCodeEditText.startAnimation(validAnimation);
```
Comment (polok) [inline library/src/main/java/pl/rspective/voucherify/android/view/VoucherCheckoutView.java]: I think that we should check if animation is already not running

File: library/src/main/res/anim/invalid.xml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<set xmlns:android="http://schemas.android.com/apk/res/android">
+    <translate
+        android:fromXDelta="-5%"
+        android:toXDelta="5%"
+        android:repeatCount="5"
+        android:repeatMode="reverse"
+        android:duration="150"/>
```
Comment (polok) [inline library/src/main/res/anim/invalid.xml:7]: Maybe use some already defined in android - https://developer.android.com/reference/android/R.integer.html#config_shortAnimTime

File: library/src/main/res/anim/valid.xml [hunk 4]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="utf-8"?>
+<set xmlns:android="http://schemas.android.com/apk/res/android">
+
+    <scale
+        android:duration="250"
```
Comment (polok) [inline library/src/main/res/anim/valid.xml:5]: As above: https://developer.android.com/reference/android/R.integer.html#config_shortAnimTime

File: library/src/main/res/anim/valid.xml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        android:pivotX="50%"
+        android:pivotY="50%"
+        android:toXScale="1.1"
+        android:toYScale="1.1"
+        android:interpolator="@android:anim/accelerate_interpolator">
+    </scale>
+
+    <scale
+        android:startOffset="250"
+        android:duration="250"
```
Comment (polok) [inline library/src/main/res/anim/valid.xml:17]: As above: https://developer.android.com/reference/android/R.integer.html#config_shortAnimTime

File: library/src/main/res/layout/voucher_checkout_view.xml [hunk 6]
```diff
@@ -25,6 +29,6 @@
             android:layout_height="wrap_content"
             android:layout_marginTop="10dp"
             android:enabled="false"
-            android:textColor="@android:color/white" />
+            style="@style/VoucherValidateButton" />
     </LinearLayout>
 </merge>
```
Comment (polok) [inline library/src/main/res/layout/voucher_checkout_view.xml:34]: Can we change the name of the file to start from 'view_' so we will have view_voucher_checkout

File: library/src/main/res/drawable/validate_button.xml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (polok) [inline library/src/main/res/drawable/validate_button.xml]: Please, add prefix 'selector' at start so we would have selector_validate_button

File: library/src/main/res/layout/voucher_checkout_view.xml [hunk 8]
```diff
@@ -7,14 +7,18 @@
         android:orientation="vertical">
 
         <android.support.design.widget.TextInputLayout
+            android:id="@+id/til_voucher_code"
             android:layout_width="match_parent"
-            android:layout_height="wrap_content">
+            android:layout_height="wrap_content"
+            android:hint="voucher code"
```
Comment (polok) [inline library/src/main/res/layout/voucher_checkout_view.xml]: Extract this one to strings xml resource file

File: library/src/main/res/drawable/validate_button.xml [hunk 9]
```diff
@@ -0,0 +1,5 @@
...
+<selector xmlns:android="http://schemas.android.com/apk/res/android">
+    <item android:state_enabled="false" android:drawable="@color/validate_button_disabled" />
+    <item android:state_enabled="true" android:drawable="@color/validate_button_enabled"/>
+</selector>
```
Comment (polok) [inline library/src/main/res/drawable/validate_button.xml]: Please add selector for 'pressed' state as well.

File: library/src/main/java/pl/rspective/voucherify/android/view/VoucherCheckoutView.java [hunk 10]
```diff
@@ -28,23 +35,34 @@
 
     private VoucherifyAndroidClient voucherifyClient;
     private OnValidatedListener onValidatedListener;
+    private TextInputLayout voucherCodeLabel;
```
Comment (polok) [inline library/src/main/java/pl/rspective/voucherify/android/view/VoucherCheckoutView.java:35]: I think that we could add option to provide 'error' message but we will think about this later. Just want to point it out

File: library/src/main/java/pl/rspective/voucherify/android/view/VoucherCheckoutView.java [hunk 11]
```diff
@@ -66,38 +85,75 @@ public void onClick(View v) {
...
+        }
+
+        String validateButtonText = a.getString(R.styleable.VoucherCheckoutView_validateButtonText);
         if (validateButtonText != null) {
             validateButton.setText(validateButtonText);
         }
 
+        voucherIcon = a.getDrawable(R.styleable.VoucherCheckoutView_voucherIcon);
+        if (voucherIcon == null) {
+            voucherIcon = ContextCompat.getDrawable(context, R.drawable.ic_voucherify_logo_white_24dp);
```
Comment (polok) [inline library/src/main/java/pl/rspective/voucherify/android/view/VoucherCheckoutView.java:105]: Please extract this code into method and apply for two validVoucherIcon and invalidVoucherIcon as well.

File: README.md [hunk 12]
```diff
@@ -191,10 +191,95 @@ VoucherResponse
...
+VoucherifyAndroidClient voucherifyClient = new VoucherifyAndroidClient.Builder(
+            YOUR-PUBLIC-CLIENT-APPLICATION-ID,
+            YOUR-PUBLIC-CLIENT-APPLICATION-TOKEN)
+       .withCustomTrackingId(YOUR-CUSTOM-TRACKNG-ID)
+       .build();
+
+VoucherCheckoutView voucherCheckout = (VoucherCheckoutView) findViewById(R.id.voucher_checkout);
+```
+
+You will also likely want to get validation results. You can achieve that by adding OnValiadtedListener:
```
Comment (lkulig) [inline README.md]: You have a typo `OnVali[ad]tedListener` it should be `OnVali[da]tedListener`.

File: library/src/main/java/pl/rspective/voucherify/android/view/VoucherCheckoutView.java (hunk 1 above)
Comment (LukaszWiktor) [inline library/src/main/java/pl/rspective/voucherify/android/view/VoucherCheckoutView.java]: Removed. Wasn't used anywhere.

File: library/src/main/res/anim/valid.xml (hunk 4 above)
Comment (LukaszWiktor) [inline library/src/main/res/anim/valid.xml:5]: What is the advantage of using the constant? I tried `shortAnimTime` but it was too fast and then `mediumAnimTime` was too slow. For this animation 250ms is the best duration time selected empirically ;)

File: library/src/main/res/layout/voucher_checkout_view.xml (hunk 6 above)
Comment (LukaszWiktor) [inline library/src/main/res/layout/voucher_checkout_view.xml:34]: Done.

File: README.md (hunk 12 above)
Comment (LukaszWiktor) [inline README.md]: @lkulig Fixed. Thanks for spotting it!

**Comment by LukaszWiktor** [comment]
@polok, @lkulig CR fixes applied

**Comment by phajduk** [comment]
LGTM :+1:  Rebase&squash it.

File: library/src/main/res/anim/valid.xml (hunk 4 above)
Comment (polok) [inline library/src/main/res/anim/valid.xml:5]: Ok, this was just a suggestion, so no worry ;)

**Comment by polok** [comment]
@LukaszWiktor I am going to merge this one but in next release we have to improve the animation as it cuts an animated view.

**Pull request merged**